cluster-autoscaler/cloudprovider/packet/packet_manager_rest.go

```diff
@@ -229,132 +250,186 @@ func randString8() string {
-			panic(resp)
+			return createDevice(cr, mgr.baseURL)
 		}
-	}
-	defer resp.Body.Close()
-
-	rbody, err := ioutil.ReadAll(resp.Body)
-	klog.V(3).Infof("Response body: %v", string(rbody))
-	if err != nil {
-		klog.Errorf("Failed to read response body: %v", err)
-	} else {
-		klog.Infof("Created new node on Packet.")
-	}
-	if cr.HardwareReservationID != "" {
-		klog.Infof("Reservation %v", cr.HardwareReservationID)
-	}
-}
 
-// createNodes provisions new nodes on packet and bootstraps them in the cluster.
-func (mgr *packetManagerRest) createNodes(nodegroup string, nodes int) error {
-	klog.Infof("Updating node count to %d for nodegroup %s", nodes, nodegroup)
-	cloudinit, err := base64.StdEncoding.DecodeString(mgr.cloudinit)
-	if err != nil {
-		log.Fatal(err)
-		return fmt.Errorf("Could not decode cloudinit script: %v", err)
-	}
-
-	for i := 0; i < nodes; i++ {
-		mgr.createNode(string(cloudinit), nodegroup)
+		return fmt.Errorf("failed to create device: %w", err)
 	}
 
 	return nil
 }
 
-func createDevice(cr *DeviceCreateRequest, baseURL string) (*http.Response, error) {
+// TODO: find a better way than parsing the error messages for this.
+func isNoAvailableReservationsError(err error) bool {
+	return strings.Contains(err.Error(), " no available hardware reservations ")
+}
+
+func createDevice(cr *DeviceCreateRequest, baseURL string) error {
 	packetAuthToken := os.Getenv("PACKET_AUTH_TOKEN")
 	url := baseURL + "/projects/" + cr.ProjectID + "/devices"
-	jsonValue, _ := json.Marshal(cr)
+
+	jsonValue, err := json.Marshal(cr)
+	if err != nil {
+		return fmt.Errorf("failed to marshal create request: %w", err)
+	}
+
 	klog.Infof("Creating new node")
 	klog.V(3).Infof("POST %s \n%v", url, string(jsonValue))
+
 	req, err := http.NewRequest("POST", url, bytes.NewBuffer(jsonValue))
 	if err != nil {
-		klog.Errorf("Failed to create device: %v", err)
-		panic(err)
+		return fmt.Errorf("failed to create request: %w", err)
 	}
+
 	req.Header.Set("X-Auth-Token", packetAuthToken)
 	req.Header.Set("Content-Type", "application/json")
 	client := &http.Client{}
+
 	resp, err := client.Do(req)
-	return resp, err
+	if err != nil {
+		return fmt.Errorf("failed to perform request: %w", err)
+	}
+
+	defer func() {
+		if err := resp.Body.Close(); err != nil {
+			klog.Errorf("failed to close response body: %v", err)
+		}
+	}()
+
+	rbody, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return fmt.Errorf("failed to read response body: %w", err)
+	}
+
+	klog.V(3).Infof("Response body: %v", string(rbody))
+
+	if resp.StatusCode == http.StatusOK || resp.StatusCode == http.StatusCreated {
+		return nil
+	}
+
+	var result interface{}
+	if err := json.Unmarshal(rbody, &result); err != nil {
+		return fmt.Errorf("request failed: %v, failed to unmarshal the response: %v, error: %w", resp, result, err)
+	}
+
+	mappedResult := result.(map[string]interface{})
+	if errorMessages, ok := mappedResult["errors"]; ok {
```

Comment:
In some conditions, the error response is singular:
```
$ curl -H "Content-Type: application/json" -H "X-Auth-Token: invalid" -X GET "https://api.packet.net/facilities" 
{"error":"Invalid credentials"}
```

packngo handles it this way:

https://github.com/packethost/packngo/blob/master/packngo.go#L257-L267
https://github.com/packethost/packngo/pull/42